File: detection-engine/internal/rules/brute_force.go
package rules

import (
	"fmt"
	"strings"
	"time"
)

// BruteForceRule детектирует brute-force атаки на API / SignalR auth.
// Sigma: sigma-rules/brute_force_api.yaml
// MITRE: T1110, T1110.001 (Brute Force: Password Guessing)
//
// Логика: COUNT(failed_auth) >= threshold за window по одному source_ip.
// Состояние хранится в Redis, ключ: "bf:{source_ip}" TTL=window.
type BruteForceRule struct {
	Threshold int           // количество неудачных попыток
	Window    time.Duration // временное окно
}

func NewBruteForceRule() *BruteForceRule {
	return &BruteForceRule{
		Threshold: 10,
		Window:    2 * time.Minute,
	}
}

func (r *BruteForceRule) ID() string    { return "brute_force_api" }
func (r *BruteForceRule) Title() string { return "API / SignalR Brute-Force Authentication Attempts" }

// Match фильтрует события — возвращает событие-кандидат если оно подходит
// под паттерн brute-force (401/403 на auth endpoints).
// Сам алерт генерируется в Evaluate после накопления счётчика.
func (r *BruteForceRule) Match(event *Event) *Alert {
	// Фильтр 1: только HTTP события с кодом ошибки авторизации
	if event.StatusCode == nil {
		return nil
	}
	code := *event.StatusCode
	if code != 401 && code != 403 {
		return nil
	}

	// Фильтр 2: auth endpoints
	path := StrVal(event.URLPath)
	authPaths := []string{"/api/auth", "/api/login", "/api/token", "/hubs/", "/api/account"}
	isAuthPath := false
	for _, p := range authPaths {
		if strings.Contains(path, p) {
			isAuthPath = true
			break
		}
	}
	if !isAuthPath {
		return nil
	}

	// Фильтр 3: должен быть source_ip
	if event.SourceIP == nil {
		return nil
	}

	return nil // алерт генерируется в Evaluate
}

// Evaluate обновляет счётчик в Redis и возвращает алерт при превышении порога.
func (r *BruteForceRule) Evaluate(event *Event, state StateStore) *Alert {
	if event.SourceIP == nil {
		return nil
	}

	// Проверяем кандидата (повторяем фильтр из Match)
	if event.StatusCode == nil {
		return nil
	}
	code := *event.StatusCode
	if code != 401 && code != 403 {
		return nil
	}
	path := StrVal(event.URLPath)
	authPaths := []string{"/api/auth", "/api/login", "/api/token", "/hubs/", "/api/account"}
	isAuthPath := false
	for _, p := range authPaths {
		if strings.Contains(path, p) {
			isAuthPath = true
			break
		}
	}
	if !isAuthPath {
		return nil
	}

	key := fmt.Sprintf("bf:%s", *event.SourceIP)
	count, err := state.Increment(key, r.Window)
	if err != nil {
		return nil
	}

	// Алерт только при точном пересечении порога (чтобы не спамить)
	if count != int64(r.Threshold) {
		return nil
	}

	ip := *event.SourceIP
	return &Alert{
		RuleID:    r.ID(),
		RuleTitle: r.Title(),
		Severity:  SeverityHigh,
		Description: fmt.Sprintf(
			"Brute-force detected: %d failed authentication attempts in %v from %s",
			r.Threshold, r.Window, ip,
		),
		SourceIP:  &ip,
		UserID:    event.UserID,
		EventIDs:  []string{event.EventID},
		MitreTags: []string{"T1110", "T1110.001"},
		FiredAt:   time.Now().UTC(),
		Context: map[string]any{
			"failed_attempts": count,
			"window":         r.Window.String(),
			"url_path":       path,
		},
	}
}

File: detection-engine-rs/src/rules/brute_force.rs
use std::collections::HashMap;
use std::time::Duration;

use async_trait::async_trait;
use chrono::Utc;

use crate::alert::{Alert, AlertSeverity};
use crate::event::Event;
use crate::state_store::StateStore;

use super::{Rule, StatefulRule, format_duration};

const AUTH_PATHS: &[&str] = &[
    "/api/auth",
    "/api/login",
    "/api/token",
    "/hubs/",
    "/api/account",
];

pub struct BruteForceRule {
    pub threshold: i64,
    pub window: Duration,
}

impl BruteForceRule {
    pub fn new() -> Self {
        Self {
            threshold: 10,
            window: Duration::from_secs(120),
        }
    }

    fn is_candidate(&self, event: &Event) -> bool {
        match event.status_code {
            Some(401) | Some(403) => {}
            _ => return false,
        }
        let path = Event::str_val(&event.url_path);
        if !AUTH_PATHS.iter().any(|p| path.contains(p)) {
            return false;
        }
        event.source_ip.is_some()
    }
}

impl Rule for BruteForceRule {
    fn id(&self) -> &str {
        "brute_force_api"
    }
    fn title(&self) -> &str {
        "API / SignalR Brute-Force Authentication Attempts"
    }
    fn match_event(&self, _event: &Event) -> Option<Alert> {
        None
    }
}

#[async_trait]
impl StatefulRule for BruteForceRule {
    async fn evaluate(&self, event: &Event, state: &dyn StateStore) -> Option<Alert> {
        if !self.is_candidate(event) {
            return None;
        }

        let ip = event.source_ip.as_ref()?;
        let key = format!("bf:{}", ip);
        let count = state.increment(&key, self.window).await.ok()?;

        if count != self.threshold {
            return None;
        }

        let path = Event::str_val(&event.url_path);
        let mut context = HashMap::new();
        context.insert("failed_attempts".into(), serde_json::json!(count));
        context.insert(
            "window".into(),
            serde_json::json!(format_duration(self.window)),
        );
        context.insert("url_path".into(), serde_json::json!(path));

        Some(Alert {
            rule_id: self.id().into(),
            rule_title: self.title().into(),
            severity: AlertSeverity::High,
            description: format!(
                "Brute-force detected: {} failed authentication attempts in {} from {}",
                self.threshold,
                format_duration(self.window),
                ip,
            ),
            source_ip: Some(ip.clone()),
            user_id: event.user_id.clone(),
            event_ids: vec![event.event_id.clone()],
            mitre_tags: vec!["T1110".into(), "T1110.001".into()],
            fired_at: Utc::now(),
            context,
        })
    }
}

#[cfg(test)]
mod tests {
    use std::sync::Arc;

    use super::*;
    use crate::rules::test_utils::{MockStateStore, event_with};

    #[tokio::test]
    async fn fires_when_threshold_is_reached_on_auth_failures() {
        let rule = BruteForceRule {
            threshold: 3,
            window: Duration::from_secs(120),
        };
        let store = Arc::new(MockStateStore::default());
        let event = event_with(|e| {
            e.status_code = Some(401);
            e.url_path = Some("/api/auth/login".into());
            e.source_ip = Some("192.168.0.10".into());
        });

        assert!(rule.evaluate(&event, store.as_ref()).await.is_none());
        assert!(rule.evaluate(&event, store.as_ref()).await.is_none());

        let alert = rule
            .evaluate(&event, store.as_ref())
            .await
            .expect("expected alert on threshold");
        assert_eq!(alert.rule_id, "brute_force_api");
        assert_eq!(alert.severity, AlertSeverity::High);
    }

    #[tokio::test]
    async fn ignores_non_auth_or_non_failure_events() {
        let rule = BruteForceRule::new();
        let store = Arc::new(MockStateStore::default());
        let event = event_with(|e| {
            e.status_code = Some(200);
            e.url_path = Some("/api/orders".into());
        });

        let alert = rule.evaluate(&event, store.as_ref()).await;
        assert!(alert.is_none());
    }
}
